File: src/utils/wasm_fingerprint.rs
use anyhow::Result;
use bs58;
use ripemd::Ripemd160;
use sha2::{Digest, Sha256};
use std::path::Path;

/// Calculates a fingerprint for a WASM module using the following algorithm:
/// 1. Compute RIPEMD-160 hash of the WASM binary
/// 2. Prepend custom type prefix (0x17)
/// 3. Compute double SHA-256 checksum
/// 4. Append checksum to the prefixed bytes
/// 5. Encode in XRP Ledger Base58 alphabet
///
/// # Arguments
///
/// * `wasm_path` - Path to the WASM file
///
/// # Returns
///
/// * `Result<String>` - The fingerprint string if successful, error otherwise
pub fn calculate_wasm_fingerprint(wasm_path: &Path) -> Result<String> {
    // Read WASM file
    let wasm_bytes = std::fs::read(wasm_path)
        .map_err(|e| anyhow::anyhow!("Failed to read WASM file: {:?}", e))?;

    // Compute RIPEMD-160 hash
    let mut ripemd = Ripemd160::new();
    ripemd.update(&wasm_bytes);
    let ripemd_hash = ripemd.finalize(); // 20 bytes

    // Prepend custom type prefix (0x17) which ensures the result starts with "w" (for 20-byte payload)
    let mut prefixed = vec![0x17];
    prefixed.extend_from_slice(&ripemd_hash);

    // Compute double SHA-256 checksum
    let first_hash = Sha256::digest(&prefixed);
    let second_hash = Sha256::digest(&first_hash);
    let checksum = &second_hash[..4];

    // Append checksum to the prefixed bytes
    let mut with_checksum = prefixed;
    with_checksum.extend_from_slice(checksum);

    // Use XRP Ledger Base58 alphabet
    let xrp_alphabet =
        bs58::Alphabet::new(b"rpshnaf39wBUDNEGHJKLM4PQRST7VWXYZ2bcdeCg65jkm8oFqi1tuvAxyz")
            .expect("Provided alphabet is invalid");

    // Encode in Base58
    let fingerprint = bs58::encode(&with_checksum)
        .with_alphabet(&xrp_alphabet)
        .into_string();

    Ok(fingerprint)
}

File: reference/js/wasm-fingerprint/wasm-fingerprint.js
#!/usr/bin/env node

const fs = require('fs');
const crypto = require('crypto');
const rippleAddressCodec = require('ripple-address-codec');

function generateFingerprint(wasmBinary) {
    // 1. Compute RIPEMD160
    const hash = crypto.createHash('ripemd160')
        .update(wasmBinary)
        .digest();
    
    // 2. Use all 20 bytes
    const fingerprintBytes = hash;
    
    // 3. Prepend WASM type prefix (0x17)
    const prefix = Buffer.from([0x17]);
    const prefixed = Buffer.concat([prefix, fingerprintBytes]);
    
    // 4. Compute double SHA-256 checksum
    const firstHash = crypto.createHash('sha256')
        .update(prefixed)
        .digest();
    const secondHash = crypto.createHash('sha256')
        .update(firstHash)
        .digest();
    const checksum = secondHash.slice(0, 4);
    
    // 5. Append checksum
    const withChecksum = Buffer.concat([prefixed, checksum]);
    
    // 6. Encode using XRPL's Base58 alphabet via ripple-address-codec
    // The ripple-address-codec uses the XRPL alphabet by default
    const fingerprint = rippleAddressCodec.encode(withChecksum, {
        // No version needed since we already prepended our own prefix
        version: null,
        // Use the ripple alphabet (default)
        alphabet: 'ripple'
    });
    
    return fingerprint;
}

function hexToBuffer(hexString) {
    // Remove any whitespace or 0x prefix and convert to lowercase
    const cleanHex = hexString.replace(/[\s0x]/g, '').toLowerCase();
    
    // Check if the hex string is valid
    if (!/^[0-9a-f]*$/.test(cleanHex)) {
        throw new Error('Invalid hex string');
    }
    
    // Check if the hex string has an even length
    if (cleanHex.length % 2 !== 0) {
        throw new Error(`Hex string must have an even number of characters (got ${cleanHex.length})`);
    }
    
    // Convert hex to buffer
    return Buffer.from(cleanHex, 'hex');
}

function main() {
    if (process.argv.length !== 3) {
        console.error('Usage: node wasm-fingerprint.js <wasm-file-path-or-hex-string>');
        process.exit(1);
    }

    const input = process.argv[2];
    let wasmBinary;

    try {
        // Check if input is a file path
        if (fs.existsSync(input) && !input.match(/^[0-9a-fA-F]+$/)) {
            wasmBinary = fs.readFileSync(input);
        } else {
            // Assume it's a hex string
            wasmBinary = hexToBuffer(input);
        }

        const fingerprint = generateFingerprint(wasmBinary);
        console.log(fingerprint);
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

main(); 
